### DISCUSSION TITLE
A weird(?) behavior of `shapes_dr`

### DISCUSSION THREAD

**Discussion by Mephisto405** [mitsuba-renderer/mitsuba3#637] [Q&A] [answered]
In the below code, I guess snippet (1) and snippet (2) both work exactly the same way, by the definition of `ShapePtr`, which is not the case here.
More specifically, snippet (1) raises an error `DiscreteDistribution: no probability mass found!`
Indeed, I intentionally tried to remove these shapes, which raise the error, by the try-except statements. Thus snippet (2) works properly without the error.
Any help would be appreciated!! 🙏🏻🙏🏻🙏🏻

For reproducing the problem:
```python
import drjit as dr
import mitsuba as mi
import torch

mi.set_variant("cuda_ad_rgb")

"""
wget https://benedikt-bitterli.me/resources/mitsuba/staircase.zip -P data/
unzip "data/*.zip" -d data/
rm data/*.zip
"""
scene = mi.load_file("/YOURPATH/data/staircase/scene_v3.xml")

shapes = scene.shapes_dr()
idx = []
for i, shape in enumerate(scene.shapes()):
    try:
        ps = shape.sample_position(0.5, [0.5, 0.5]).p.torch()
        print(i, ps.mean(), shape.surface_area())
        idx.append(i)
    except:
        continue
shapes_reduced = dr.gather(mi.ShapePtr, scene.shapes_dr(), idx)

"""Snippet (1) This does not work"""
shapes_reduced.sample_position(0.5, [0.5, 0.5])

"""Snippet (2) This does work"""
# for i, j in zip(range(len(shapes_reduced)), idx):
#     print(j)
#     shapes_reduced[i].sample_position(0.5, [0.5, 0.5])
```

**Comment by njroussel** [comment, marked as answer]
Hi @Mephisto405 

Basically `shapes.reduces.sample_position` is a virtual function call. Unless you disable vcall recording with `dr.set_flag(dr.JitFlag.VCallRecord, False)`, Dr.Jit will "trace/read" through **every** possible target function and then assemble a single optimized kernel which is finally executed.

In your case, even though some target function will not actually be executed, it is still recorded and during the recording an error happens. I've looked at the error myself, it seems that some of the meshes in the scene are "broken". For example `Mesh432.obj` is just a line of vertices.

You have 3 options:
* Do a non-parallel evaluation as you suggest in snippet 2
* Disable virtual function recording (locally around the scope of your `sample_position` call)
* Fix the scene such that it doesn't have any broken meshes
